pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2ServiceProviderMetadataResolver.java

```diff
@@ -129,17 +132,27 @@ private void writeServiceProviderMetadataToResource(final String tempMetadata) t
         final WritableResource metadataResource = configuration.getServiceProviderMetadataResource();
         if (metadataResource != null && CommonHelper.isNotBlank(tempMetadata)) {
             if (metadataResource.exists() && !configuration.isForceServiceProviderMetadataGeneration()) {
-                logger.info("Metadata file already exists at {}.", metadataResource.getFile());
+                logger.info("Metadata file already exists at {}.", metadataResource.getFilename());
             } else {
                 logger.info("Writing sp metadata to {}", metadataResource.getFilename());
-                final File parent = metadataResource.getFile().getParentFile();
-                if (parent != null) {
-                    logger.info("Attempting to create directory structure for: {}", parent.getCanonicalPath());
-                    if (!parent.exists() && !parent.mkdirs()) {
-                        logger.warn("Could not construct the directory structure for SP metadata: {}",
-                            parent.getCanonicalPath());
+                File resourceFile;
+                try {
+                    resourceFile = metadataResource.getFile();
+                    if (resourceFile != null) {
+                        final File parent = resourceFile.getParentFile();
+                        if (parent != null) {
+                            logger.info("Attempting to create directory structure for: {}", parent.getCanonicalPath());
+                            if (!parent.exists() && !parent.mkdirs()) {
+                                logger.warn("Could not construct the directory structure for SP metadata: {}",
+                                       parent.getCanonicalPath());
+                            }   
+                        }
                     }
+                } catch (UnsupportedOperationException e) {
+                    // do nothing since likely a resource that doesn't have a filesystem
+                    logger.warn("no filesystem", e);
                 }
```

Comment:
This entire block seems like it should be moved to the file-based resolver.

Reply:
It is not 100% certain that particular resolver does or does not support directories. You can't know this ahead of time. Plus the current file-based resolver isn't pac4j code.

Reply:
Understood. Either way, it has to move out of there and go somewhere else. It was semi sensible when we were explicitly using the file-backed resolver; Now that it's abstracted out, the responsibility does not seem like it would belong here. This code is clearly dealing with the file system; so either the code has to move out and find a new home, or ideally, it would have to go into the file-backed resolver (or our extension of it that would own this bit)

Reply:
But the issue is that not only filesystems support directory structures as is evident in the abstraction. An object store may do the same. But I can attempt to move it and see if it makes sense